File: CommandNavigation/CommandStack.cs
namespace CommandNavigation {
  using System;
  using System.Collections.Generic;
  using System.Text;

  public class CommandStack<T> : Stack<T> {

    /// <summary>
    /// 
    /// </summary>
    /// <param name="OnPushed"></param>
    /// <param name="OnPopped"></param>
    /// <param name="Detecting">left:Top item in stack;right:item need push;return:True->pop the top one</param>
    public CommandStack(Action<T> OnPushed, Action<T> OnPopped, Func<T, T, bool> Detecting) : base() {
      _OnPushed = OnPushed ?? (E => { });
      _OnPopped = OnPopped ?? (E => { });
      _Detecting = Detecting ?? ((L, R) => false);
    }
    /// <summary>
    /// 
    /// </summary>
    /// <param name="OnPushed"></param>
    /// <param name="OnPopped"></param>
    /// <param name="Detecting">left:Top item in stack;right:item need push;return:True->pop the top one</param>
    /// <param name="Capacity"></param>
    public CommandStack(Action<T> OnPushed, Action<T> OnPopped, Func<T, T, bool> Detecting, int Capacity) : base(Capacity) {
      _OnPushed = OnPushed ?? (E => { });
      _OnPopped = OnPopped ?? (E => { });
      _Detecting = Detecting ?? ((L, R) => false);
    }

    protected readonly Action<T> _OnPushed;
    protected readonly Action<T> _OnPopped;
    protected readonly Func<T, T, bool> _Detecting;

    public new void Push(T Item) {
      if (Count == 0) {
        base.Push(Item);
        _OnPushed(Item);
      }
      else if (_Detecting(Peek(), Item)) {
        _OnPopped(base.Pop());
        Push(Item);
      }
      else {
        base.Push(Item);
        _OnPushed(Item);
      }
    }

    public new T Pop() {
      var Item = base.Pop();
      _OnPopped(Item);
      return Item;
    }

    public new void Clear() {
      while (Count > 0) Pop();
    }

    public void Discard() {
      base.Clear();
    }
  }
}

File: CommandNavigation.TS/CommandStack.ts

import { Func2, Event2 } from "./../KatKits/KatKits.TS/Event";

export default class CommandStack<T> extends Array<T>{
  public readonly OnPushed: Event2<CommandStack<T>, T> = new Event2<CommandStack<T>, T>();
  public readonly OnPopped: Event2<CommandStack<T>, T> = new Event2<CommandStack<T>, T>();

  protected readonly ComparerDelegate: Func2<T, T, boolean>;

  constructor(Comparer: Func2<T, T, boolean>) {
    super();
    this.ComparerDelegate = Comparer;
  }
  public Peek(): T { return this[this.length - 1]; }
  public Push(Item: T): void {
    if (this.length == 0) {
      super.push(Item);
      this.OnPushed.Invoke(this, Item);
    }
    else if (this.ComparerDelegate(this.Peek(), Item)) {
      this.OnPopped.Invoke(this, super.pop());
      this.Push(Item);
    }
    else {
      super.push(Item);
      this.OnPushed.Invoke(this, Item);
    }
  }

  public Pop(): T {
    var Item = super.pop();
    this.OnPopped.Invoke(this, Item);
    return Item;
  }

  public Clear() {
    while (this.length > 0) {
      this.Pop();
    }
  }
  public Discard() { super.splice(0, this.length); }

}
